designs/RFC - Network Setup Teardown Optimization.md

```diff
@@ -0,0 +1,396 @@
+    %8 = comp.schedule_read %arg2 from %4 on %0 wait for %7 : (memref<16x32xf32>, memref<16x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    %9 = comp.schedule_read %arg1 from %5 on %0 wait for %7 : (memref<8x16xf32>, memref<8x16xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    %10 = comp.schedule_read %arg3 from %1 on %0 wait for %7 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    comp.wait %8, %9, %10 : !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>
+  }
+  comp.destroy_kernel %k1 : !comp.kernel<ocl, 11, 8x32xf32>
+  comp.destroy_kernel %k2 : !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.dealloc %0 %4 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>) -> ()
+  comp.dealloc %0 %5 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>) -> ()
+  comp.dealloc %0 %1 : (!comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>) -> ()
+  comp.destroy_execenv %0 : !comp.execenv<ocl:0,(11)>
+  return
+}
+```
+
+### Final Lowering
+
+With a minor tweak, we could run the above code as a coroutine &mdash; we could pass in an extra parameter to represent
+a bidirectional request/response pipe, and pass it to `comp.loop`.  Our sense is that it's more obvious to split the
+current `main()` function into multiple functions invoked individually to set up the network, run it, and tear it down.
+
+To do this, we augment the existing `!comp.execenv` to serve as a resource container, and add accessor operations for
+the various resources.  For our example code, this transformation looks something like this:
+
+```mlir
+func @plaidml_init(%arg0: !comp.device) -> !comp.execenv<ocl:0,11)> {
+  %0 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
+  %1 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<8x32xf32, 11>
+  comp.execenv_set_alloc %0, %1 {midx = 0} : !comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>
+  %4 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<16x32xf32, 11>
+  comp.execenv_set_alloc %0, %4 {midx = 1} : !comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>
+  %5 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<8x16xf32, 11>
+  comp.execenv_set_alloc %0, %5 {midx = 2} : !comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>
+  %k1 = comp.create_kernel %0 {kernel = @main_kernel::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 8x32xf32>
+  comp.execenv_set_kernel %0, %k1 {kidx = 0} : !comp.execenv<ocl:0,(11)>, !comp.kernel<ocl, 11, 8x32xf32>
+  %k2 = comp.create_kernel %0 {kernel = @main_kernel_0::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.execenv_set_kernel %0, %k2 {kidx = 1} : !comp.execenv<ocl:0,(11)>, !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
```

Comment:
I think we should return these values instead of 'setting' them onto the environment. We'll then need to inject these values into the parameters of `plaidml_exec`

Reply:
Hmm... that's tricky; what do you lower to?  (I mean, what's the ABI even look like?)  I have another idea for how we could do this, though, by creating another type we can lower to an LLVMIR struct -- just a sec, I'll write it up.

Reply:
I think we can pass allocations as `memref` directly, and I think we can add a new type for kernels.

Reply:
I think the lowering for kernels could actually turn into an `int32` or whatnot, but that should only be at the LLVM dialect level I would think.

Reply:
I think we should avoid using LLVM structs unless we have a real good reason to use one.

Reply:
PTAL